### PR TITLE
Control setup

### PR DESCRIPTION
**Pull request by AdamuKaapan**
# Goals
* Implement the basic plumbing of the control system (minus actual hardware interface, that'll come later)
* Create an accurate-enough model of the robot for use in simulation/testing based on CAD measurements

# Works Completed
For the hardware interface itself, I copied and then heavily modified the same files from the gazebo prototype `control` branch, where I was testing these concepts earlier. This resulted in the tfr_control files controller_launcher and controller (launcher is a node, controller is just a class), as well as the controllers.yaml file. I also created a launch file to test this controller setup in tfr_control.

I also set up the model itself, which resulted in four files in tfr_description/xacro: model.xacro, and 3 other files with the model_ prefix (to distinguish from Collin's work earlier with the proto_ prefix). With this I set up a launch file in tfr_visualization (view_model) which displays the robot in rviz. Please note that the two front wheels being misplaced with no transform is intentional and expected.

# Results
The plumbing of the control system is now set up entirely and is ready to have hardware/simulation wired into it and can be used by other packages for plumbing setup. We also now have a model which is close enough to our modeled measurements to be used effectively for MoveIt testing.

The work on this pull request has also opened my eyes to a potential for testing MoveIt without using Gazebo and using only rviz. This will be brought up at an upcoming meeting and is a subject for a later pull request.

# Testing
There's two launch files that can be used for testing. To test the hardware interface, run tfr_control's test_controller_launch.launch. You can verify that this is working by opening rqt_graph and look for a set of topics for every controller (grouping by namespace helps). To test the model file, run tfr_visualization's view_model.launch. Verify that all the joints are controllable correctly by moving the sliders on the window that appears (it might be behind rviz, check on the taskbar) and checking that the correct joints move and the range of motion appears correct.

### CODE REVIEW

**Comment by LinuxKid21** [comment]
Hey there. I tried running "roslaunch tfr_control test_controller_launcher.launch" and got some errors. Here is the full output:

xeno@xeno-ThinkPad-11e:~/NasaRmc2018$ roslaunch tfr_control test_controller_launcher.launch 
... logging to /home/xeno/.ros/log/2fd924c2-f5c2-11e7-a91e-4ceb42e6586b/roslaunch-xeno-ThinkPad-11e-10906.log
Checking log directory for disk usage. This may take awhile.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://xeno-ThinkPad-11e:44339/

SUMMARY
========

PARAMETERS
 * /arm_controller/gains/lower_arm_joint/d: 0.0
 * /arm_controller/gains/lower_arm_joint/i: 500.0
 * /arm_controller/gains/lower_arm_joint/p: 1.0
 * /arm_controller/gains/turntable_joint/d: 0.0
 * /arm_controller/gains/turntable_joint/i: 500.0
 * /arm_controller/gains/turntable_joint/p: 1.0
 * /arm_controller/gains/upper_arm_joint/d: 0.0
 * /arm_controller/gains/upper_arm_joint/i: 500.0
 * /arm_controller/gains/upper_arm_joint/p: 1.0
 * /arm_controller/joints: ['turntable_joint...
 * /arm_controller/type: effort_controller...
 * /arm_end_controller/gains/scoop_joint/d: 0.0
 * /arm_end_controller/gains/scoop_joint/i: 500.0
 * /arm_end_controller/gains/scoop_joint/p: 1.0
 * /arm_end_controller/joints: ['scoop_joint']
 * /arm_end_controller/type: effort_controller...
 * /bin_position_controller/joint: bin_joint
 * /bin_position_controller/pid/d: 0.0
 * /bin_position_controller/pid/i: 500.0
 * /bin_position_controller/pid/p: 1.0
 * /bin_position_controller/type: effort_controller...
 * /joint_state_controller/publish_rate: 50
 * /joint_state_controller/type: joint_state_contr...
 * /left_tread_velocity_controller/joint: left_tread_joint
 * /left_tread_velocity_controller/pid/d: 0.0
 * /left_tread_velocity_controller/pid/i: 500.0
 * /left_tread_velocity_controller/pid/p: 1.0
 * /left_tread_velocity_controller/type: effort_controller...
 * /right_tread_velocity_controller/joint: right_tread_joint
 * /right_tread_velocity_controller/pid/d: 0.0
 * /right_tread_velocity_controller/pid/i: 500.0
 * /right_tread_velocity_controller/pid/p: 1.0
 * /right_tread_velocity_controller/type: effort_controller...
 * /robot_description: <?xml version="1....
 * /rosdistro: kinetic
 * /rosversion: 1.12.12

NODES
  /
    controller_launcher (tfr_control/controller_launcher)
    controller_spawner (controller_manager/spawner)
    robot_state_publisher (robot_state_publisher/robot_state_publisher)

ROS_MASTER_URI=http://localhost:11311

process[robot_state_publisher-1]: started with pid [10926]
process[controller_launcher-2]: started with pid [10927]
process[controller_spawner-3]: started with pid [10930]
[WARN] [1515560010.827026]: DEPRECATION warning: --shutdown-timeout has no effect.
[ERROR] [1515560010.855441265]: Could not load controller 'joint_state_controller' because controller type 'joint_state_controller/JointStateController' does not exist.
[ERROR] [1515560010.855593803]: Use 'rosservice call controller_manager/list_controller_types' to get the available types
[ERROR] [1515560011.857933]: Failed to load joint_state_controller
[ERROR] [1515560011.873180150]: Could not load controller 'left_tread_velocity_controller' because controller type 'effort_controllers/JointVelocityController' does not exist.
[ERROR] [1515560011.873271053]: Use 'rosservice call controller_manager/list_controller_types' to get the available types
[ERROR] [1515560012.875737]: Failed to load left_tread_velocity_controller
[ERROR] [1515560012.893798966]: Could not load controller 'right_tread_velocity_controller' because controller type 'effort_controllers/JointVelocityController' does not exist.
[ERROR] [1515560012.894151251]: Use 'rosservice call controller_manager/list_controller_types' to get the available types
[ERROR] [1515560013.897370]: Failed to load right_tread_velocity_controller
[ERROR] [1515560013.911407862]: Could not load controller 'bin_position_controller' because controller type 'effort_controllers/JointPositionController' does not exist.
[ERROR] [1515560013.911863652]: Use 'rosservice call controller_manager/list_controller_types' to get the available types
[ERROR] [1515560014.916260]: Failed to load bin_position_controller
[ERROR] [1515560014.931700675]: Could not load controller 'arm_controller' because controller type 'effort_controllers/JointTrajectoryController' does not exist.
[ERROR] [1515560014.931782206]: Use 'rosservice call controller_manager/list_controller_types' to get the available types
[ERROR] [1515560015.934183]: Failed to load arm_controller
[ERROR] [1515560015.949886074]: Could not load controller 'arm_end_controller' because controller type 'effort_controllers/JointTrajectoryController' does not exist.
[ERROR] [1515560015.950041216]: Use 'rosservice call controller_manager/list_controller_types' to get the available types
[ERROR] [1515560016.952590]: Failed to load arm_end_controller

**Comment by AdamuKaapan** [comment]
@LinuxKid21 Hmm... it builds fine on the build server from a fresh install where I installed `ros-kinetic-desktop-full`. The first thing that comes to mind is for you to `sudo apt-get install` these packages: ros-kinetic-controller-manager, ros-kinetic-hardware-interface, ros-kinetic-effort-controllers, and ros-kinetic-joint-state-controller. If this doesn't work lemme know if the error changes and I'll think of more things to try.

**Comment by AdamuKaapan** [comment]
Sorry, just realized I was stupid, and building and launching aren't the same thing. Regardless, try those steps anyways. I'll go through the build server again and test some steps and see if I get the same issue.

**Comment by AdamuKaapan** [comment]
@LinuxKid21 Okay, after testing on the build server (a completely fresh install), I've found that installing these packages works: ros-kinetic-joint-state-controller, ros-kinetic-effort-controllers, and ros-kinetic-joint-trajectory-controller. Can you confirm this works for you?

**Comment by AdamuKaapan** [comment]
@collinsc I've marked some extra packages as dependencies. Could you please review to see if I've done this correctly? You're more familiar with ROS dependency management than I am.

**Review by LinuxKid21 — DISMISSED** [review]
Alright. Everything seems good on my end now. The new changes make it so it fails to compile if I don't have the packages installed so I'd say it's working quite fine.

File: src/tfr_control/config/controllers.yaml
Comment (collinsc) [inline src/tfr_control/config/controllers.yaml:20]: Have you tuned these PID values using the tuning tutorial on gazebo?  They seem a little wacky.

It doesn't need to be done for the gazebo model at all, but we do need to do it for the real thing.

File: src/tfr_control/include/tfr_control/controller.h
Comment (collinsc) [inline src/tfr_control/include/tfr_control/controller.h]: you got some uninitialized arrays yo! Fix this.
double command_values[7] = {};

Comment (collinsc) [reply src/tfr_control/include/tfr_control/controller.h]: This is an issue because they are not in global scope and unspecified.

File: src/tfr_control/package.xml
Comment (collinsc) [inline src/tfr_control/package.xml:9]: This is a seperate pull request but in the future we should decide on a team email to put down for the maintainer contact.

File: src/tfr_control/src/controller.cpp
Comment (collinsc) [inline src/tfr_control/src/controller.cpp:12]: typically method implementations in a cpp file match the order they are declared in the header, fix me.

File: src/tfr_control/src/controller_launcher.cpp
Comment (collinsc) [inline src/tfr_control/src/controller_launcher.cpp:27]: probably because controller manager is threaded and multiplexes between multiple controllers.

File: src/tfr_description/xacro/model_arm.xacro
Comment (collinsc) [inline src/tfr_description/xacro/model_arm.xacro:125]: Are we giving any of these components inertia?

Comment (collinsc) [reply src/tfr_description/xacro/model_arm.xacro:125]: doesnt need to be this pull request, but this might screw with move it.

**Comment by collinsc** [comment]
I have a deprecation warning @AdamuKaapan do you know where that comes from?
[WARN] [1515607722.724648]: DEPRECATION warning: --shutdown-timeout has no effect.
image

**Comment by AdamuKaapan** [comment]
@collinsc that literally just now started appearing on mine too, I hadn't seen it ever before until yesterday. I'll take a look once I'm home tonight.

**Review by collinsc — CHANGES_REQUESTED** [review]
Launches look good, I raise some questions for future pull requests. However your c++ code has some initialization issues that could bite you later, and one bit of style that needs to be improved.

File: src/tfr_control/include/tfr_control/controller.h
Comment (LinuxKid21) [reply src/tfr_control/include/tfr_control/controller.h]: What could they be initialized with? Maybe I'm wrong here, but don't they have to initialized by the controller_manager with their real physical values? If so, then any values given to them will not be any better than whatever random stuff was in memory.

Maybe we could create a variable that tracks if they have been initialized with write, that throws an error if they have not been written to yet?

Comment (collinsc) [reply src/tfr_control/include/tfr_control/controller.h]: Thats a little complicated. The brace will initialize them to 0. If we want a different value we could do:

Arr my Art[]={-1}
Its a one line fix,and initializing arrays is a best practice in c++11.

Keep simple things simple

Comment (LinuxKid21) [reply src/tfr_control/include/tfr_control/controller.h]: I'm fine with the change. I'm just not sure it actually helps (zero doesn't seem any better than random stuff in memory if both are wrong).

So I think the best thing to do would be initializing them to something that is obviously wrong so the program doesn't "accidentally work" sometimes and fail at others randomly.

Comment (RyanBerge) [reply src/tfr_control/include/tfr_control/controller.h]: I agree with Levi.  Initializing variables to 0 or some other default value is best practice because sometimes people assume incorrectly that 0 *is* the default value, and use them as though that were the case.  Here, *any* usage of the values before they've been initialized by the ROS framework is unacceptable.

I would recommend adding an `initialized` flag and then putting `assert` statements in `read` and `write` to check the flag.  Or we could throw exceptions in those functions instead, since ROS makes heavy use of exceptions anyway.

**Review by RyanBerge — CHANGES_REQUESTED** [review]
I put in my two cents on the array initialization subject.  Most of my feedback I gave you in person, but you should also change the name of `test_controller_launcher.launch` to remove the `_launcher` suffix, per coding standard specification, especially since you also have a node called `controller_launcher.cpp`.

In addition, I'd like you to edit the `README.md` file to add an "Installation" header that describes what additional packages need to be installed from ROS in order to run your code should the pull request be approved.  In this case, the `libsfml-dev` package for the graphics thing (that might have been from Collin, but we still need it).  While you're at it, if you could add the GTest installation/setup instructions, that would be excellent.

File: src/tfr_control/include/tfr_control/controller.h
Comment (AdamuKaapan) [reply src/tfr_control/include/tfr_control/controller.h]: I think I'd prefer initializing with zero vs any other value as that seems more likely to result in less random hardware movement (if something were to happen perchance). Adding a flag also seems definitely doable, I'll experiment with `assert` vs exceptions vs something else (if it exists) and get back to you on that.

File: src/tfr_control/package.xml
Comment (AdamuKaapan) [reply src/tfr_control/package.xml:9]: Agreed

File: src/tfr_control/src/controller_launcher.cpp
Comment (AdamuKaapan) [reply src/tfr_control/src/controller_launcher.cpp:27]: That and I think cm.update actually hangs on waiting for a topic update or something, so it needs a spinner on another thread to do it?

File: src/tfr_description/xacro/model_arm.xacro
Comment (AdamuKaapan) [reply src/tfr_description/xacro/model_arm.xacro:125]: Currently I don't have any, but for setting up Gazebo for Moya I'll need to. That's a separate pull request though, yeah. As far as screwing with MoveIt, I guess we'll find out.

**Comment by AdamuKaapan** [comment]
Thanks all! I'll get to all these changes either tonight (after 20 pages of CSS 350 reading, so meh probably not) or tomorrow morning.

File: src/tfr_control/include/tfr_control/controller.h
Comment (collinsc) [reply src/tfr_control/include/tfr_control/controller.h]: @RyanBerge @LinuxKid2 @AdamuKaapan  If we are going to make any sort of assumption about that, then let's just leave it. I did not intend for this to become any sort of contested issue. My thing is if I see an uninitialized variable in code I have been assigned to review. I'll mark it, it's on my list of things that should be done. 

I think this exception thing is excessive, it should just be  a quick fix and move on.

File: src/tfr_description/xacro/model_arm.xacro
Comment (collinsc) [reply src/tfr_description/xacro/model_arm.xacro:125]: I did a little bit of that already, there are a few kinks but check out that proto model I checked in.

**Comment by collinsc** [comment]
@RyanBerge @AdamuKaapan  I think if there is any readme editing, that sounds like a seperate pull request to me. That sounds like a job.

**Comment by collinsc** [comment]
I also think having 4 reviewers might be overkill adam, I think we are tripping all over eachother, and it's making the conversation tend more toward bikeshedding than towards actual review. Hence the initialization conversation.

File: src/tfr_control/include/tfr_control/controller.h
Comment (AdamuKaapan) [reply src/tfr_control/include/tfr_control/controller.h]: No worries. Initializing it I don't think will be an issue anyways, I'll go ahead and do that, I don't think we need to worry about checking it.

**Comment by AdamuKaapan** [comment]
@collinsc that's true. Quick explanation of how that happened, I added Brad and Levi because they're assigned to review my stuff, I added Ryan because he's also working on control stuff, and I added you when the dependency issue came up, I agree it's a bit much. I'll keep that in mind for my next PR. If I have one specific thing that I want to check up on from someone not in the PR, should I just ping them in Slack and have them look at it real quick, instead of making them a full reviewer?

For the readme, I agree, that might be a separate request.

**Comment by collinsc** [comment]
I just think our pull requests will get bogged down with too many people. We all mean well, but when it gets to the point where we are nitpicking and correcting each other and other _reviewers_ on established c++ best practices, that's getting out of hand in my book. 

Let's just pick something and move on, we have a lot of code to write.

File: src/tfr_control/config/controllers.yaml
Comment (AdamuKaapan) [reply src/tfr_control/config/controllers.yaml:20]: Nope, these aren't tweaked yet. We'll be tuning them once we have physical hardware, or possibly sooner for the gazebo model depending on if it needs it.

**Comment by AdamuKaapan** [comment]
@collinsc fixed the declaration order issue in controller.cpp, also initialized the arrays to zero but didn't add any assertions/checks. @RyanBerge renamed the launch file to remove "_launcher".

**Review by LinuxKid21 — DISMISSED** [review]

**Review by RyanBerge — DISMISSED** [review]

**Review by collinsc — APPROVED** [review]
Looks Good! Let's make sure to keep note of the inertia values and maintainer email for the future.

**Review by RyanBerge — APPROVED** [review]
It was a little rough going, but that's okay.  We had a pull request open for weeks when I was at UASC once.  It had hundreds of comments by the time it was merged.

You did good work, and it's even better now.  That's the whole point of code reviews.

**Pull request merged**
